src/Simplex/RemoteControl/Invitation.hs

```diff
@@ -174,3 +157,67 @@ requiredP q k f = maybe (fail $ "missing " <> show k) (either fail pure . f) $ l
 -- optionalP q k f = maybe (pure Nothing) (either fail (pure . Just) . f) $ lookup k q
 
 $(JQ.deriveJSON defaultJSON ''RCInvitation)
+
+sessionAddressJSON :: RCInvitation -> ByteString
+sessionAddressJSON = LB.toStrict . J.encode . J.toJSON
+
+signInviteJSON :: C.PrivateKey C.Ed25519 -> C.PrivateKey C.Ed25519 -> RCInvitation -> ByteString
+signInviteJSON sKey idKey invitation = idSigned
+  where
+    sessionAddress = sessionAddressJSON invitation
+
+    ssig :: C.Signature 'C.Ed25519
+    ssig =
+      case C.sign (C.APrivateSignKey C.SEd25519 sKey) sessionAddress of
+        C.ASignature C.SEd25519 s -> s
+        _ -> error "signing with ed25519"
+    sSigned = sessionAddress <> C.signatureBytes ssig
+
+    idsig :: C.Signature 'C.Ed25519
+    idsig =
+      case C.sign (C.APrivateSignKey C.SEd25519 idKey) sSigned of
+        C.ASignature C.SEd25519 s -> s
+        _ -> error "signing with ed25519"
+    idSigned = sSigned <> C.signatureBytes idsig
+
+instance Encoding RCSignedInvitation where
+  smpEncode RCSignedInvitation {invitation, ssig, idsig} = sessionAddressJSON invitation <> C.signatureBytes ssig <> C.signatureBytes idsig
```

Comment:
this is brittle and should instead have length prefixes, so it should be something like smpEncode (sessionAddressJSON invitation, ssig, idsig)

Signatures know how to encode.

Then your decoder will be as simple:

(json, ssig, idsig) <- smpP